bsdtar: Support `--mtime` and `--clamp-mtime`

zhaofengli opened the pull request:
Hi,

This PR adds support for setting a forced mtime on all written files (`--mtime` and `--clamp-mtime`) in bsdtar.

The end goal will be to support all functionalities in <https://reproducible-builds.org/docs/archives/#full-example>, namely `--sort` and disabling other attributes (atime, ctime, etc.).

Fixes #971.

## History

- [v1](https://github.com/zhaofengli/libarchive/tree/forced-mtime-v1): Added `archive_read_disk_set_forced_mtime` in libarchive. As a result, it was only applied when reading from the filesystem and not from other archives.
- [v2](https://github.com/zhaofengli/libarchive/tree/forced-mtime-v2): Refactored to apply the forced mtime in `archive_write`.
- v3 (current): Reduced libarchive change to exposing `archive_parse_date`, moved clamping logic into bsdtar.

DHowett commented on libarchive/archive_read_disk_posix.c:
```diff
@@ -1244,6 +1250,38 @@ archive_read_disk_set_matching(struct archive *_a, struct archive *_ma,
+}
+
+int
+archive_read_disk_set_forced_mtime_str(struct archive *_a,
```
Keeping this in `..._posix` will likely prevent bsdtar from linking properly on Windows. Is this feature intended to be specific to non-Windows platforms? It doesn't look platform-specific. 🙂

DHowett reviewed (changes requested):
Setting aside my concerns about the platform-specificity of this work, I wonder if overriding the mtime at _read time_ is truly ideal.

`tar` supports creating an archive from another archive (`bsdtar -c -f foo.tar @bar.cpio`); should `--mtime` be applied on `disk` and only during `read`, it would be impossible to override the mtime on archive-to-archive copies.

zhaofengli commented:
> tar supports creating an archive from another archive (bsdtar -c -f foo.tar @bar.cpio); should --mtime be applied on disk

Oh, I wasn't aware and missed this while skimming the source. bsdtar keeps surprising me with its versatility! Let me figure out what the best place is to apply this (and future reproducibility-related configs/transformations).

zhaofengli commented:
Ok, I've refactored it to apply the forced mtime on `archive_write` instead and updated the tests. The `--mtime` is now applied on archive-to-archive copies as well.

kientzle commented:
> If we want to generalize, we could have some internal archive_write_entry_filter abstraction for transformations that mutate file metadata before writing.

No.  This does not belong in libarchive.

Part of libarchive's design is for consumers to request entries one at a time and then pass them along.  That gives you the option to mutate the entries in any way you see fit.  There's no need for the complexity of a modular system of "entry filters."

kientzle commented on libarchive/archive.h:
```diff
@@ -796,6 +796,25 @@ __LA_DECL int archive_write_get_bytes_in_last_block(struct archive *);
+ * If clamp is not 0, the given timestamp will only be applied if the file
+ * is newer than it.
+ */
+__LA_DECL int	archive_write_set_forced_mtime(struct archive *,
```
I really don't like this approach.   I don't think any changes to libarchive are needed; everything should be handled in bsdtar.

kientzle commented:
Please take a close look at how the existing UID, GID, uname, gname overrides are handled in the functions [`write_heirarchy()`](https://github.com/libarchive/libarchive/blob/a612bf62f86a6faa47bd57c52b94849f0a404d8c/tar/write.c#L885) in tar/write.c.   You can extend this logic to also support overriding the mtime or other attributes as needed.

The equivalent logic for archive-to-archive copies is in [`append_archive()`](https://github.com/libarchive/libarchive/blob/a612bf62f86a6faa47bd57c52b94849f0a404d8c/tar/write.c#L672), which seems to currently lack support for UID, GID, uname, and gname overrides.

I suggest you change this PR to not modify libarchive at all.  Instead focus on the two functions above in bsdtar.  I would first submit a PR that adds the mtime override logic to `write_heirarchy()` and a test for same.  Then in a separate PR, please look for ways to refactor this code to unify the entry edits so that `append_archive()` supports all the same operations.

kientzle commented:
For reproducible builds, it might be worth instead pursuing a `--reproducible` option that simply strips all data from the entry except for the filename, file type, and file size.  Everything else can be left unset, which will result in default values being used for any data that is mandatory for a particular format.   That would leave sorting (which is likely going to require changes to libarchive itself) as the only remaining requirement.

zhaofengli commented:
> I suggest you change this PR to not modify libarchive at all.

Ok, so it's permissible to do something like the following `bsdtar`?

```c
#define __LIBARCHIVE_BUILD 1
#include "archive_getdate.h"
#define get_date __archive_get_date
```

zhaofengli commented:
Re-read my previous reply and realized I may have come off as terse. This wasn't my intention, sorry. To expand, I thought about my options (I didn't like adding to libarchive itself initially, too):

1. Expose `archive_get_date`, a generic date string parser, as public API. This comes off as weird for an archiving library. It makes sense to take a date string specifically for matching (archive_match), but less so for general use.
2. Access the private `__archive_get_date` API from bsdtar. Since bsdtar is just a consumer and can be configured to build against the system libarchive, I don't think this is what we want to do.
3. Follow how it was done with `archive_match` (`--newer`/`--older`) which takes either a `time_t` or a libarchive-specific date string directly for matching. This is how this PR is currently implemented.

To me (3) made the most sense, but I'm happy to change to whatever approach you prefer.

kientzle commented:
Your reasoning makes sense, but I'm not happy with the implication:  This suggests that we should put a lot of clamping logic (default and overridden UID, GID, gname, uname, ctime, atime, file modes, etc) into libarchive.  And even then, how would you support users who wanted, for example, to override the file permissions for directories but not regular files?  This is basically the reason that libarchive does not have a single "extract this archive" function.   By requiring the client to iterate over the entries, the client can do any filtering they might want outside of libarchive.

I'd honestly be happier either exposing the existing date parser (though perhaps under a better name -- maybe `archive_parse_date`) or even copying that one source file into the tar sources.

kientzle commented:
(I had missed the dependence on date parsing ... thank you for pointing that out.)

emaste commented:
> it might be worth instead pursuing a `--reproducible` option

That may well make sense to do, but even if so we should still support `--mtime` and --clamp-mtime` for GNU tar compatibility.

There are also cases where we want the created tarball to both be reproducible and have meaningful/sensible timestamps that are not 1970-01-01 or so. For instance, we can use the timestamp of the most recent src tree commit when creating the tarball of the FreeBSD src tree included in the release artifacts.

zhaofengli commented:
I've refactored so the only change to `libarchive` now is to expose `archive_parse_date`. bsdtar uses this function to parse `--mtime` and applies the timestamp in `append_archive` and `write_hierarchy` (we can unify this later). For the `--mtime` test I now test archive-to-archive copying as well. I've opted not to copy the date parsing code so the behavior of `--newer`/`--older` and `--mtime` remain consistent.

The best way to review the diff is going commit-by-commit.

kientzle reviewed (approved):
Looks great!

zhaofengli commented:
Interesting failure on Windows. The relevant line for the first `tar -cf` appears to be:

> The system cannot find the path specified.

Sadly I don't have Windows build environment at hand at the moment. Will try to set it up tomorrow or over the weekend.

<details>
<summary>Full log</summary>

```
Reference files will be read from: D:/a/libarchive/libarchive/tar/test
Running tests on: "D:\a\libarchive\libarchive\build_ci\cmake\bin\bsdtar.exe"
Exercising: bsdtar 3.8.0 - libarchive 3.8.0dev zlib/1.3 liblzma/5.6.3 bz2lib/1.1.0 libzstd/1.5.6 cng/1.0 libb2/bundled 
The system cannot find the path specified.
bsdtar.exe: Error opening archive: Failed to open '../noclamp.tar'
The system cannot find the path specified.
bsdtar.exe: Error opening archive: Failed to open '../clamp.tar'
The system cannot find the path specified.
bsdtar.exe: Error opening archive: Failed to open '../archive2archive.tar'

 51: test_option_mtime
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:23: 0 != systemf("%s --format pax -cf ../noclamp.tar " "--mtime \"1970/1/2 0:0:0 UTC\" .", testprog)
      0=0 (0x0, 00)
      systemf("%s --format pax -cf ../noclamp.tar " "--mtime \"1970/1/2 0:0:0 UTC\" .", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:31: 0 != systemf("%s xf ../noclamp.tar", testprog)
      0=0 (0x0, 00)
      systemf("%s xf ../noclamp.tar", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:32: Can't access new_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:33: Can't access mid_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:34: Can't access old_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:39: 0 != systemf("%s --format pax -cf ../clamp.tar " "--mtime \"1970/1/2 0:0:0 UTC\" --clamp-mtime .", testprog)
      0=0 (0x0, 00)
      systemf("%s --format pax -cf ../clamp.tar " "--mtime \"1970/1/2 0:0:0 UTC\" --clamp-mtime .", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:47: 0 != systemf("%s xf ../clamp.tar", testprog)
      0=0 (0x0, 00)
      systemf("%s xf ../clamp.tar", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:48: Can't access new_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:49: Can't access mid_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:50: Can't access old_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:54: 0 != systemf("%s --format pax -cf ./archive2archive.tar " "--mtime \"1970/1/1 0:0:0 UTC\" @noclamp.tar", testprog)
      0=0 (0x0, 00)
      systemf("%s --format pax -cf ./archive2archive.tar " "--mtime \"1970/1/1 0:0:0 UTC\" @noclamp.tar", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:60: 0 != systemf("%s xf ../archive2archive.tar", testprog)
      0=0 (0x0, 00)
      systemf("%s xf ../archive2archive.tar", testprog)=1 (0x1, 01)
D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:61: Can't access new_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:62: Can't access mid_mtime

D:\a\libarchive\libarchive\tar\test\test_option_mtime.c:63: Can't access old_mtime


Totals:
  Tests run:                1
  Tests failed:             1
  Assertions checked:      40
  Assertions failed:       15
  Skips reported:           0

Failing tests:
  51: test_option_mtime (15 failures)
```
</details>

DHowett commented:
I have a Windows build environment at hand. Let me take a look. :)

DHowett commented:
The `system` API on Windows is, unfortunately, a loaded footgun.

```patch
From 9b85c944a4cce96e3dc4ff1ac63fdb8602e74b4d Mon Sep 17 00:00:00 2001
From: "Dustin L. Howett" <dustin@howett.net>
Date: Wed, 14 May 2025 21:25:01 -0500
Subject: [PATCH] Fix the tests on Windows

---
 tar/test/test_option_mtime.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/tar/test/test_option_mtime.c b/tar/test/test_option_mtime.c
index 8e337a46..75cdb839 100644
--- a/tar/test/test_option_mtime.c
+++ b/tar/test/test_option_mtime.c
@@ -6,6 +6,23 @@
  */
 #include "test.h"
 
+#if defined(_WIN32) && !defined(__CYGWIN__)
+/* system() on Windows runs its arguments through CMD.EXE, which has
+ * notoriously unfriendly quoting rules. The current best documented way around
+ * them is to wrap your *entire commandline* in sacrificial quotes.
+ *
+ * See CMD.EXE /? for more information. Excerpted here:
+ * | Otherwise, old behavior is to see if the first character is
+ * | a quote character and if so, strip the leading character and
+ * | remove the last quote character on the command line, preserving
+ * | any text after the last quote character.
+ *
+ * Since this test makes heavy use of systemf() with quoted arguments inside
+ * the commandline, this macro is unfortunately an easier workaround.
+ */
+#define systemf(command, ...) systemf("\"" command "\"", __VA_ARGS__)
+#endif
+
 DEFINE_TEST(test_option_mtime)
 {
 	/* Create three files with different mtimes. */
-- 
2.47.1.windows.1
```

zhaofengli commented:
> The system API on Windows is, unfortunately, a loaded footgun.

Thanks for investigating! I wonder if using single quotes would have made it easier, but at least it works now :)

> GNU tar reports a failure if you specify `--clamp-mtime` with no `--mtime`:
> 
> ```
> tar: --clamp-mtime needs a date specified using --mtime
> <exited with code 2>
> ```
> 
> Should bsdtar do the same?

Good catch, I've added a check. While we're at it, I've also updated the help text and the manpage.

zhaofengli commented:
Thanks for merging! Just a note - I see from the commit log that you typically use GitHub's "Squash and Merge" feature. This can make the history hard to follow. For example, here "Expose archive_parse_date as a public API" and "Add support for --mtime and --clamp-mtime" are logically distinct steps, but now the small code changes for the actual feature are mixed together with the noisy refactoring.

In the future, perhaps you can consider using "Rebase and Merge" instead which preserves the history:

<img src="https://github.com/user-attachments/assets/99a802a6-aaf8-4920-a0fd-210593dd69f9" width=500></img>
